doc/sphinxext/redirect_from.py

```diff
@@ -48,32 +49,61 @@
 def setup(app):
     RedirectFrom.app = app
     app.add_directive("redirect-from", RedirectFrom)
+    app.add_domain(RedirectFromDomain)
     app.connect("build-finished", _generate_redirects)
 
+    metadata = {'parallel_read_safe': True}
+    return metadata
+
+
+class RedirectFromDomain(Domain):
+    """
+    The sole purpose of this domain is a parallel_read_safe data store for the
+    redirects mapping.
+    """
+    name = 'redirect_from'
+    label = 'redirect_from'
+
+    @property
+    def redirects(self):
+        """The mapping of the redirectes."""
+        return self.data.setdefault('redirects', {})
+
+    def clear_doc(self, docnames):
+        self.redirects.clear()
+
+    def merge_domaindata(self, docnames, otherdata):
+        for src, dst in otherdata['redirects'].items():
+            if src not in self.redirects:
+                self.redirects[src] = dst
+            elif self.redirects[src] != dst:
+                raise ValueError(
+                    f"{src} is already noted as redirecting to {dst}")
```

Comment:
`dst` is from `otherdata`, so it's not _already_ redirecting to it, but to `self.redirects[src]`. Best to just show both URLs.